### PR TITLE
Transform knative-operator to use go modules

### PR DESCRIPTION
**Pull request by markusthoemmes**
Let's see how this goes. This is blocking us from being more consistent with our dependencies.

**Note:** This contains a "manual" patch for manifestival, that I expect to be temporary until we can properly upgrade everything to more recent versions. The transform to go modules is kinda needed as a prereq for that though.

### CODE REVIEW

**Comment by markusthoemmes** [comment]
/test 4.5-upgrade-tests-aws-ocp-45

Infra

**Comment by markusthoemmes** [comment]
/hold 

until our KnativeKafka work lands.

**Review by rhuss — COMMENTED** [review]
Looks good to me in general, and if it passes the tests then we should merge anyway.

However, reviewing this PR is hard. Would be cool if there would be one commit for the changes in vendor only, and another for the 'real change's.

**Comment by markusthoemmes** [comment]
Ack @rhuss, I'll pull that apart for a proper review.

**Comment by markusthoemmes** [comment]
@rhuss done, all real changes are now in commit 1.

File: knative-operator/hack/update-deps.sh
Comment (rhuss) [reply knative-operator/hack/update-deps.sh:13]: So iiur, the patch is needed because the upstream manifestival only works with new k8s versions. But why are we need not pinning the manifestival also to an older version that works with k8s 1.13 instead of patching it on the fly ?

File: knative-operator/hack/update-deps.sh
Comment (rhuss) [inline knative-operator/hack/update-deps.sh:20]: Upstream script also removes LICENSE files. 

Other remarks:

* You could combine to a single find 
* This command will break totally if files names contains spaces. So user `xargs` instead.
* It also breaks if there are too many files so that the overall command line becomes too long.
* Why not deleting just all `*_test.go` files ?

Comment (rhuss) [reply knative-operator/hack/update-deps.sh:20]: Recommended form to fix those issues:

```
find "vendor/" \( -name "OWNERS" -o -name "*_test.go" \) -print0 | xargs -0 rm -f
```

**Review by rhuss — COMMENTED** [review]
Thanks for splitting the commits. Find my comments inline.

File: knative-operator/go.mod
Comment (markusthoemmes) [reply knative-operator/go.mod:19]: They are both overridden in the replace section to nail them to what we currently have to use.

File: knative-operator/go.mod
Comment (markusthoemmes) [reply knative-operator/go.mod:23]: 0.17.3 is used, but we cannot update currently because we get clashes with client-go dependencies that we in turn cannot update because we have to use an old controller-runtime because of the way our webhooks are setup.

I plan to untangle all this (this PR is keeping all dependencies roughly the same as they used to be before) to be able to update everything to much more appropriate versions.

File: knative-operator/go.mod
Comment (markusthoemmes) [reply knative-operator/go.mod:30]: I haven't deeply checked, but I guess in some apiserver dependency. I had issues with these upstream too. The dep indeed is super old and didn't get updated in between.

File: knative-operator/go.mod
Comment (markusthoemmes) [reply knative-operator/go.mod:38]: Some other dependency we have depends on a newer version of controller-runtime and thus make go mod take the newer one above.

File: knative-operator/go.mod
Comment (markusthoemmes) [reply knative-operator/go.mod:32]: See longer comment above.

File: knative-operator/hack/update-deps.sh
Comment (markusthoemmes) [reply knative-operator/hack/update-deps.sh:13]: I fighted with go modules quite a while to make this work with the older manifestival version, but it would throw errors at me. This seemed the most non-intrusive way forward for now.

File: knative-operator/hack/update-deps.sh
Comment (markusthoemmes) [reply knative-operator/hack/update-deps.sh:20]: Totally copied from upstream, lemme check.

File: knative-operator/hack/manifestival.patch
Comment (matzew) [inline knative-operator/hack/manifestival.patch:4]: Is it worth to apply upstream? and eventually use a release containing the patch ?

Comment (rhuss) [reply knative-operator/hack/manifestival.patch:4]: I think as soon as we get the k8s deps straight, we get rid of this patch anyway. It's more a 'backwards' patch (to make manifestival compatible with older k8s), so don't think upstream is keen on downgrading their k8s deps :)

Comment (rhuss) [reply knative-operator/hack/manifestival.patch:4]: (this is how I understoo @markusthoemmes )

File: knative-operator/hack/update-deps.sh
Comment (rhuss) [inline knative-operator/hack/update-deps.sh:16]: 👍

Comment (rhuss) [reply knative-operator/hack/update-deps.sh:16]: btw, is the file mode change on the scripts intended, so that they are not executable anymore  ? (see the other files in this commit, can't really comment on them directly)

Happens because you have removed the chmod line.

File: knative-operator/hack/manifestival.patch
Comment (markusthoemmes) [reply knative-operator/hack/manifestival.patch:4]: Exactly, we only need this because of our hilariously old k8s deps.

File: knative-operator/hack/update-deps.sh
Comment (markusthoemmes) [reply knative-operator/hack/update-deps.sh:16]: For now that seems fine, yes. We can always bring that back should we actually need to execute files.

**Review by rhuss — APPROVED** [review]
Looks good for me, please unhold when ready to merge

/hold
/lgtm
/approve

File: knative-operator/go.mod
Comment (aliok) [reply knative-operator/go.mod:23]: Also, IMO, keeping the versions the same with `Gopkg.toml` is a better idea. ...unless we really really need to change them here because of some `go mod` problems.

Actually upgrading the dependency versions should be in a separate PR.

Comment (markusthoemmes) [reply knative-operator/go.mod:23]: Agree @aliok, that was the goal of this.

**Comment by aliok** [comment]
/lgtm
/approve

Thanks for doing this.

But, let's keep `/hold` for a while until we're done with big Kafka related PRs

**Comment by aliok** [comment]
/unhold

we're done with big Kafka PRs. We can merge this.

**Pull request merged**
